nodego/supervisor.go

```diff
@@ -0,0 +1,189 @@
+	currentLogBatch = &logBatch{
+		ready: make(chan struct{}),
+	}
+	logBatchesToReport <- currentLogBatch
+	return currentLogBatch
+}
+
+type supervisorWriter string
+
+const isoTimeFormat = "2006-01-02T15:04:05.999Z07:00"
+
+func (w supervisorWriter) Write(p []byte) (n int, err error) {
+	entry := &logEntry{
+		TextPayload: string(p),
+		Severity:    string(w),
+		Time:        time.Now().Format(isoTimeFormat),
+	}
+	if functionExecutionId != "" {
+		entry.ExecutionID = functionExecutionId
+	}
+	currentLogBatchMutex.Lock()
+	defer currentLogBatchMutex.Unlock()
+	if currentLogBatch == nil {
+		return os.Stderr.Write(entry.ConsoleOutput())
+	}
+	// Start a new batch if the current one would grow too much.
+	if len(currentLogBatch.Entries)+1 > maxLogBatchEntries || currentLogBatch.payloadLength+len(entry.TextPayload) > maxLogBatchLength {
+		startNewLogBatch()
+	}
+	currentLogBatch.AddEntry(entry)
+	return len(entry.TextPayload), nil
+}
+
+func logBatchReportWorker() {
+	for logBatch := range logBatchesToReport {
+		<-logBatch.ready
+
+		currentLogBatchMutex.Lock()
+		if logBatch == currentLogBatch {
+			startNewLogBatch()
+		}
+		currentLogBatchMutex.Unlock()
+
+		// If the next batch is a regular non-empty log batch, report it.
+		postData, _ := json.Marshal(logBatch)
+		if err := postToSupervisor("/_ah/log", postData, supervisorLogTimeout); err != nil {
+			killInstance()
+			return
+		}
+	}
+}
+
+func postToSupervisor(path string, postData []byte, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	req, _ := http.NewRequest("POST", (&url.URL{
+		Scheme: "http",
+		Host:   supervisorHostname + ":" + supervisorInternalPort,
+		Path:   path,
+	}).String(), bytes.NewBuffer(postData))
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Content-Length", strconv.Itoa(len(postData)))
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		select {
+		case <-ctx.Done():
+			fmt.Fprintln(os.Stderr, "Timeout when calling supervisor")
+			return fmt.Errorf("Timeout when calling supervisor")
+		default:
+			fmt.Fprintf(os.Stderr, "Error when calling supervisor: %s\n\n%s\n", err.Error(), debug.Stack())
+			return fmt.Errorf("Error when calling supervisor")
+		}
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		fmt.Fprintf(os.Stderr, "Incorrect response code when calling supervisor: %d\n", resp.StatusCode)
+		return fmt.Errorf("Incorrect response code from supervisor")
+	}
+
+	return nil
+}
+
+func killInstance() {
+	postToSupervisor("/_ah/kill", nil, supervisorKillTimeout)
+	os.Exit(16)
+}
+
+// WithLogger returns an http handler that reads the function execution id and
+// attaches it to log messages sent to the supervisor.
+func WithLogger(handler http.Handler) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		functionExecutionId = r.Header.Get("Function-Execution-Id")
+		handler.ServeHTTP(w, r)
+	}
```

Comment:
If I recall correctly, worker.js sends all pending log entries when the request returns. This avoids having log lines buffered too long and avoids losing log lines if the instance is killed shortly after handling a request.

Reply:
I feel like this should not be too much of a problem unless there's more than a few full batches of logs since the worker should be sending off log batches as fast as possible.

Reply:
Here are some scenarios to consider:
1. You are testing your app. You make a request and want to see the logs. You check the logs, but none show up. You wait a few minutes, refresh again and still don't see the logs.
1. You have a low-traffic app that gets one request every few hours. You notice that you rarely get logs from it and when you do they are greatly delayed (and possibly truncated).
1. Similar to the first scenario, only in this case you only see some of the log lines instead of none of them.

In each case, the user is likely assume something is broken (most likely the unofficial Go runtime, but possibly Stackdriver Logging).

Reply:
Remember that the post-request flush does not need to be synchronous, kicking off an asynchronous task is probably best.